### PR TITLE
Deprecate on_keyboard_interrupt callback hook

### LINKED ISSUE #9286: Handling of KeyboardInterrupt
## 🐛 Bug
Same behavior as in https://github.com/PyTorchLightning/pytorch-lightning/issues/6807

May be caused due to the wrong indentation of the raise statement?
https://github.com/PyTorchLightning/pytorch-lightning/blob/75350938ca646efc0b4bac432ba2d5d4676662bb/pytorch_lightning/trainer/trainer.py#L524

### To Reproduce
Run training and raise a Keyboard Interrupt.

### Expected behavior
Training should end and KeyboardInterrupt should stop the whole program.

### Environment

- PyTorch Lightning Version (e.g., 1.3.0):
- PyTorch Version (e.g., 1.8)
- Python version:
- OS (e.g., Linux):
- CUDA/cuDNN version:
- GPU models and configuration:
- How you installed PyTorch (`conda`, `pip`, source):
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

### PR DESCRIPTION
**Pull request by daniellepintz**
## What does this PR do?

This callback hook will be subsumed by the new on_exception hook (PyTorchLightning#9183).
The PR includes small addition to the is_overridden helper function, to automatically set the parent to Callback if a Callback instance is passed, which made changes to configuration_validator.py simpler.

Also fixes #9286

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/configuration_validator.py [hunk 1]
```diff
@@ -201,3 +203,12 @@ def __check_training_step_requires_dataloader_iter(self, model: "pl.LightningMod
...
                     "is incompatible with `truncated_bptt_steps > 0`."
                 )
+
+    def _check_on_keyboard_interrupt(self) -> None:
+        """Checks if on_keyboard_interrupt is overriden and sends a deprecation warning."""
+        for callback in self.trainer.callbacks:
+            if is_overridden(method_name="on_keyboard_interrupt", instance=callback):
+                rank_zero_deprecation(
+                    "The `on_keyboard_interrupt` callback hook was deprecated in v1.5 and will be removed in v1.7."
+                    "Please use the `on_exception` callback hook instead."
```
Comment (awaelchli) [inline pytorch_lightning/trainer/configuration_validator.py:213]:
```suggestion
                    "The `on_keyboard_interrupt` callback hook was deprecated in v1.5 and will be removed in v1.7."
                    " Please use the `on_exception` callback hook instead."
```
some spacing needed after the period

File: CHANGELOG.md [hunk 2]
```diff
@@ -167,12 +167,16 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
...
 - Deprecated the `TestTubeLogger` ([#9065](https://github.com/PyTorchLightning/pytorch-lightning/pull/9065))
 
+
 - Deprecated `on_{train/val/test/predict}_dataloader()` from `LightningModule` and `LightningDataModule` [#9098](https://github.com/PyTorchLightning/pytorch-lightning/pull/9098)
 
 
 - Updated deprecation of  `argparse_utils.py` from removal in 1.4 to 2.0 ([#9162](https://github.com/PyTorchLightning/pytorch-lightning/pull/9162))
 
 
+- Deprecated `on_keyboard_interrupt` callback hook ([#9183](https://github.com/PyTorchLightning/pytorch-lightning/pull/9183))
```
Comment (awaelchli) [inline CHANGELOG.md:177]:
```suggestion
- Deprecated `on_keyboard_interrupt` callback hook in favor of new `on_exception` hook ([#9183](https://github.com/PyTorchLightning/pytorch-lightning/pull/9183))
```

**Review by awaelchli — APPROVED** [review]
LMTG

File: pytorch_lightning/callbacks/base.py [hunk 3]
```diff
@@ -264,7 +264,12 @@ def on_predict_end(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule")
         pass
 
     def on_keyboard_interrupt(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
```
Comment (Borda) [inline pytorch_lightning/callbacks/base.py:266]: Thinking we shall have better check if a user has a custom callback using deprecated hook...
So it can be something like inspection that callback has a method which was in the past valid hook?
cc: @PyTorchLightning/core-contributors

File: pytorch_lightning/trainer/configuration_validator.py [hunk 4]
```diff
@@ -43,6 +43,8 @@ def verify_loop_configurations(self, model: "pl.LightningModule") -> None:
         elif self.trainer.state.fn == TrainerFn.PREDICTING:
             self.__verify_predict_loop_configuration(model)
         self.__verify_dp_batch_transfer_support(model)
+        # TODO(@daniellepintz): Delete _check_on_keyboard_interrupt in v1.7
```
Comment (Borda) [inline pytorch_lightning/trainer/configuration_validator.py:46]:
```suggestion
        # TODO: Delete _check_on_keyboard_interrupt in v1.7
```

File: CHANGELOG.md (hunk 2 above)
Comment (ananthsub) [reply CHANGELOG.md]: @awaelchli should we add this to the contributing section?  Something like "what to do if you're deprecating something?" 

Includes how to even initiate the process, what type of sign-off or approval is needed, and then the mechanics of emitting warnings, adding tests, and making sure the change log reflects what to use instead? 

Then the governance docs around API lifecycle could remain the high-level strategy around this

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/callbacks/base.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/base.py:266]: @Borda there's a similar suggestion discussed here: https://github.com/PyTorchLightning/pytorch-lightning/discussions/9070

File: CHANGELOG.md (hunk 2 above)
Comment (daniellepintz) [reply CHANGELOG.md]: resolving to allow for merge

**Comment by daniellepintz** [comment]
Waiting to merge - will incorporate fix for #9286

**Comment by Borda** [comment]
> Waiting to merge

there are still some failing tests, mind have look?

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -505,13 +505,15 @@ def _call_and_handle_interrupt(self, trainer_fn: Callable, *args: Any, **kwargs:
...
             return trainer_fn(*args, **kwargs)
+        # TODO(@daniellepintz): treat KeyboardInterrupt as BaseException (delete the code below) in v1.7
         except KeyboardInterrupt as exception:
             rank_zero_warn("Detected KeyboardInterrupt, attempting graceful shutdown...")
             # user could press Ctrl+c many times... only shutdown once
             if not self.interrupted:
                 self.state.status = TrainerStatus.INTERRUPTED
                 self.on_keyboard_interrupt()
                 self.on_exception(exception)
+                raise exception
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:516]: @daniellepintz I would leave this change to another PR. This seemed like an intentional choice from https://github.com/PyTorchLightning/pytorch-lightning/pull/856/ to not re-raise the exception. Doing now will be a breaking change. Let's discuss this on https://github.com/PyTorchLightning/pytorch-lightning/issues/9286

File: pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -505,13 +505,15 @@ def _call_and_handle_interrupt(self, trainer_fn: Callable, *args: Any, **kwargs:
         """
         try:
             return trainer_fn(*args, **kwargs)
+        # TODO(@daniellepintz): treat KeyboardInterrupt as BaseException (delete the code below) in v1.7
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:508]: not sure about the convention around github names in the codebase. so far we've left todos without them. @Borda is there a style guide we ought to file for todos?

File: pytorch_lightning/utilities/model_helpers.py [hunk 7]
```diff
@@ -58,6 +50,16 @@ def is_overridden(
...
+    if parent is None:
+        if isinstance(instance, pl.LightningModule):
+            parent = pl.LightningModule
+        elif isinstance(instance, pl.LightningDataModule):
+            parent = pl.LightningDataModule
+        elif isinstance(instance, pl.Callback):
+            parent = pl.Callback
+        if parent is None:
+            raise ValueError("Expected a parent")
+
```
Comment (ananthsub) [inline pytorch_lightning/utilities/model_helpers.py:62]: i think this change is causing the `is_overridden` test to fail

File: pytorch_lightning/trainer/trainer.py (hunk 5 above)
Comment (daniellepintz) [reply pytorch_lightning/trainer/trainer.py:516]: sg!

File: pytorch_lightning/trainer/trainer.py (hunk 6 above)
Comment (daniellepintz) [reply pytorch_lightning/trainer/trainer.py:508]: I saw quite a few TODOs with the github names, so figured I would add mine as well so I can easily search for all of my todos in the future. happy to change though if that's not what we want

Comment (Borda) [reply pytorch_lightning/trainer/trainer.py:508]: I would not rather have names with todos, all todos shall be transferable so anyone can take it, meaning add enough information instead of writing your name (my personal experience is that I won't remember in a few weeks anyway :])

Comment (daniellepintz) [reply pytorch_lightning/trainer/trainer.py:508]: sg!

File: pytorch_lightning/trainer/trainer.py [hunk 8]
```diff
@@ -521,7 +522,7 @@ def _call_and_handle_interrupt(self, trainer_fn: Callable, *args: Any, **kwargs:
             # reset bookkeeping
             self.state.stage = None
             self.on_exception(exception)
-            raise
+            raise exception
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:525]: This change is not necessary, is it?

Comment (daniellepintz) [reply pytorch_lightning/trainer/trainer.py:525]: no it's not, I thought it was clearer this way but I can put it back

Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:525]: Yes please :)

https://stackoverflow.com/a/45986945

**Review by Borda — APPROVED** [review]
